src/zarr/api/asynchronous.py

```diff
@@ -312,7 +312,7 @@ async def open(
     store_path = await make_store_path(store, mode=mode, path=path, storage_options=storage_options)
 
     # TODO: the mode check below seems wrong!
-    if "shape" not in kwargs and mode in {"a", "r", "r+"}:
+    if "shape" not in kwargs and mode in {"a", "r", "r+", "w"}:
```

Comment:
👍 
Based on the [list of modes](https://github.com/zarr-developers/zarr-python/blob/c070940e1f6c94b658c5e78c22e0d6466ea9591d/src/zarr/core/common.py#L43), I think the only thing missing here is "w-", so in the future we could also consider simply checking that mode is not "w-", but I don't think we need to do that here.